### PR TITLE
Execute IHostingStart instances in the primary assembly

### PR DESCRIPTION
**Pull request by Tratcher**
Also make the sample app runnable.
#1000
@danroth27 

ApplicationName is set when you call UseStartup or Configure (or you can set it explicitly). We take that name and search the same assembly for HostingStartup attributes/implementations and run them.

### CODE REVIEW

**Comment by davidfowl** [comment]
Didn't you tests mroe than one attribute in the same assembly?

File: src/Microsoft.AspNetCore.Hosting/Internal/WebHostOptions.cs [hunk 1]
```diff
@@ -26,7 +26,9 @@ public WebHostOptions(IConfiguration configuration)
             Environment = configuration[WebHostDefaults.EnvironmentKey];
             WebRoot = configuration[WebHostDefaults.WebRootKey];
             ContentRootPath = configuration[WebHostDefaults.ContentRootKey];
-            HostingStartupAssemblies = configuration[WebHostDefaults.HostingStartupAssembliesKey]?.Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries) ?? new string[0];
+            // Search the primary assembly and configured assemblies.
+            HostingStartupAssemblies = (ApplicationName + ";" + configuration[WebHostDefaults.HostingStartupAssembliesKey] ?? "")
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Hosting/Internal/WebHostOptions.cs]: `$"{ApplicationName};{configuration[WebHostDefaults.HostingStartupAssembliesKey]}"`

File: samples/SampleStartups/FakeServer.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+﻿using System.Threading;
```
Comment (JunTaoLuo) [inline samples/SampleStartups/FakeServer.cs]: I think we add copyright even in samples. Same for the other file.

File: samples/SampleStartups/FakeServer.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace SampleStartups
+{
+    // We can't reference real servers in this sample without creating a circular repo dependency.
+    // This fake server lets us at least run the code.
+    public class FakeServer : IServer
+    {
+        public IFeatureCollection Features => new FeatureCollection();
+
+        public Task StartAsync<TContext>(IHttpApplication<TContext> application, CancellationToken cancellationToken)
```
Comment (JunTaoLuo) [inline samples/SampleStartups/FakeServer.cs]: nit: => Task.CompletedTask. Same below.

File: samples/SampleStartups/FakeServer.cs (hunk 2 above)
Comment (Tratcher) [inline samples/SampleStartups/FakeServer.cs]: No, but it's very inconsistent.

**Comment by Tratcher** [comment]
@davidfowl Updated with an opt-out config switch and more tests.

**Pull request merged**